File: app/src/main/java/org/cerion/tasklist/sync/Sync.java
package org.cerion.tasklist.sync;


import android.app.Activity;
import android.content.Context;
import android.util.Log;

import org.cerion.tasklist.data.AppDatabase;
import org.cerion.tasklist.data.GoogleTasksAPI;
import org.cerion.tasklist.data.IGoogleTasksAPI;
import org.cerion.tasklist.data.Task;
import org.cerion.tasklist.data.TaskDao;
import org.cerion.tasklist.data.TaskList;
import org.cerion.tasklist.data.TaskListDao;

import java.util.Date;
import java.util.List;

import androidx.annotation.Nullable;

public class Sync {

    private static final String TAG = Sync.class.getSimpleName();
    private static final int SYNC_ADD_LIST = 0;
    private static final int SYNC_CHANGE_LIST = 1;
    private static final int SYNC_DELETE_LIST = 2;
    private static final int SYNC_ADD_TASK = 3;
    private static final int SYNC_CHANGE_TASK = 4;
    private static final int SYNC_DELETE_TASK = 5;

    private static final boolean RESYNC_WEB = false;

    //Instance variables
    private GoogleTasksAPI mAPI = null;
    private TaskDao taskDb;
    private TaskListDao listDb;
    final int[] googleToDb = { 0, 0, 0, 0, 0, 0 }; //Add Change Delete Lists / Tasks
    final int[] dbToGoogle = { 0, 0, 0, 0, 0, 0 };


    /**
     * Get token and run sync process in background
     * @param context Context
     * @param activity Use for permissions prompt if starting from activity
     * @param callback Listener for when sync completes
     */
    public static void run(Context context, @Nullable Activity activity, OnSyncCompleteListener callback)
    {
        String token = AuthTools.getSavedToken(context);

        if(token != null) {
            SyncTask task = new SyncTask(context, token, callback);
            task.execute();
        } else {
            //Get new token then run sync
            AuthTools.getTokenAndSync(context, activity, callback);
        }
    }

    Sync(Context context, String sAuthKey) {
        AppDatabase db = AppDatabase.getInstance(context);
        taskDb = db.taskDao();
        listDb = db.taskListDao();
        mAPI = new GoogleTasksAPI(sAuthKey);
    }

    boolean run() throws IGoogleTasksAPI.APIException {
        List<TaskList> googleLists = mAPI.taskLists.list();
        if(googleLists.size() == 0)
            return false;

        List<TaskList> dbLists = listDb.getAll();

        //Google->Local (Added or Updated)
        for(TaskList curr : googleLists) {
            TaskList dbList = TaskList.Companion.get(dbLists,curr.getId());
            //--- UPDATE
            if(dbList != null) {
                if(!curr.getTitle().contentEquals(dbList.getTitle())) {
                    //Name mismatch, update only if local list was not renamed
                    if(!dbList.isRenamed()) {
                        listDb.update(curr);
                        googleToDb[SYNC_CHANGE_LIST]++;
                    }
                }
            }
            //--- MERGE default, first sync only
            else if(curr.isDefault()) {
                dbList = TaskList.Companion.getDefault(dbLists);
                if(dbList != null) {
                    if(!dbList.isRenamed()) {
                        dbList.setTitle(curr.getTitle());
                        listDb.update(dbList);
                    }

                    listDb.updateId(dbList.getId(), curr.getId()); //assign ID
                    dbList.setId(curr.getId());
                }
                else
                    Log.e(TAG,"missing default list");
            }
            //--- ADD
            else { //Does not exist locally, add it
                listDb.add(curr);
                googleToDb[SYNC_ADD_LIST]++;

            }
        }

        //--- DELETE
        //Verify database list still exists on web, otherwise it was deleted
        for(TaskList curr : dbLists) {
            if(curr.getHasTempId()) //Temp ids don't exist on web since they have not been created yet
                continue;

            TaskList googleList = TaskList.Companion.get(googleLists,curr.getId());
            if(googleList == null) {
                // TODO this will cascade delete so log tasks count its removing...
                listDb.delete(curr);
                googleToDb[SYNC_DELETE_LIST]++;
            }

        }

        // TODO always reload?
        //If any changes made to local database, just reload it
        if(googleToDb[SYNC_DELETE_LIST] > 0 || googleToDb[SYNC_ADD_LIST] > 0)
            dbLists = listDb.getAll();

        //Local ----> Google
        for(TaskList dbList : dbLists) {
            //--- ADD
            if(dbList.getHasTempId()) {
                TaskList addedList = mAPI.taskLists.insert(dbList);
                if(addedList != null) {
                    listDb.updateId(dbList.getId(), addedList.getId());
                    dbList.setId(addedList.getId());
                    googleLists.add(addedList);
                    dbToGoogle[SYNC_ADD_LIST]++;
                }
                else {
                    Log.e(TAG, "Failed to add list");
                    return false;
                }
            }

            //--- UPDATE
            if(dbList.isRenamed()) {
                TaskList googleList = TaskList.Companion.get(googleLists,dbList.getId());

                if (googleList != null) {
                    if(mAPI.taskLists.update(dbList)) {
                        //Save state in db to indicate rename was successful
                        dbList.setRenamed(false);
                        listDb.update(dbList);
                        dbToGoogle[SYNC_CHANGE_LIST]++;
                    }
                    else
                        Log.d(TAG,"Failed to rename list");
                }
                else {
                    //This shouldn't be possible, if list does not exist it should have deleted database list in earlier code
                    Log.e(TAG,"Error: Failed to find list to update");
                    return false;
                }

            }

            //--- DELETE LATER, allow task lists to be deleted locally, remove entry from array if success so we don't loop below
        }

        for(TaskList list : googleLists) {
            TaskList dbList = TaskList.Companion.get(dbLists, list.getId());
            if(dbList != null)
                syncTasks(list, dbList.getUpdated());
            else
                Log.e(TAG,"Unable to find database list"); //TODO throw exception
        }

        Log.d(TAG, "Google to DB: Lists (" + googleToDb[0] + "," + googleToDb[1] + "," + googleToDb[2] + ") Tasks (" + googleToDb[3] + "," + googleToDb[4] + "," + googleToDb[5] + ")");
        Log.d(TAG, "DB to Google: Lists (" + dbToGoogle[0] + "," + dbToGoogle[1] + "," + dbToGoogle[2] + ") Tasks (" + dbToGoogle[3] + "," + dbToGoogle[4] + "," + dbToGoogle[5] + ")");

        return true;
    }

    private void syncTasks(TaskList list, Date savedUpdatedNEW) throws IGoogleTasksAPI.APIException {

        if(list.getUpdated().getTime() == 0) {
            Log.e(TAG,"invalid updated time"); //TODO, need new exception for this class
            return;
        }

        Date webUpdated = list.getUpdated();
        //Date dtLastUpdated = null;
        String listId = list.getId();

        Log.d(TAG, "syncTasks() " + listId + "\t" + savedUpdatedNEW + "\t" + webUpdated);

        List<Task> webTasks = null;
        if (savedUpdatedNEW.getTime() == 0) {
            Log.d(TAG, "New list, getting all");
            webTasks = mAPI.tasks.list(listId, null);
        } else if(RESYNC_WEB) {
            Log.d(TAG, "Re-syncing web, getting all");
            webTasks = mAPI.tasks.list(listId, null);
        }
        else if (webUpdated.after(savedUpdatedNEW)) {
            //The default list can get its modified time updated without having any new tasks, we'll get 0 tasks here sometimes but not much we can do about it
            Log.d(TAG, "Getting updated Tasks");
            Log.d(TAG, "Web   = " + webUpdated);
            Log.d(TAG, "Saved = " + savedUpdatedNEW);

            //Increase by 1 second to avoid getting previous updated record which already synced
            webTasks = mAPI.tasks.list(listId, new Date(savedUpdatedNEW.getTime() + 1000)  );
        }

        /**************
         Web -> Database
         **************/
        List<Task> dbTasks = taskDb.getAllbyList(listId);
        if (webTasks != null) {
            for (Task task : webTasks) {
                Task dbTask = getTask(dbTasks, task.getId());
                if (dbTask != null) {
                    if (task.getDeleted()) {
                        taskDb.delete(task);
                        googleToDb[SYNC_DELETE_TASK]++;
                        dbTasks.remove(dbTask);
                    } else if(dbTask.getDeleted()) {
                        Log.d(TAG,"Ignoring update since deleted on local"); //Local task is deleted and will be handled on next phase
                    }
                    else {
                        //Conflict
                        if(dbTask.getUpdated().getTime() > task.getUpdated().getTime())
                            Log.e(TAG,"Conflict: Local task was updated most recently");
                        else {
                            dbTasks.remove(dbTask);
                            taskDb.update(task);
                            googleToDb[SYNC_CHANGE_TASK]++;
                        }

                    }
                } else if (task.getDeleted()) {
                    Log.d(TAG,"Ignoring web delete since record was never added locally");
                } else {
                    taskDb.add(task);
                    googleToDb[SYNC_ADD_TASK]++;
                }
            }
        }



        /**************
         Database -> Web
         **************/
        boolean bListUpdated = false;
        for (Task task : dbTasks)
        {
            //Log.d(TAG,"Title = " + task.title + "\t" + task.updated);
            boolean bModified = false;
            if(savedUpdatedNEW.getTime() == 0 || task.getUpdated().after(savedUpdatedNEW))
                bModified = true;

            if(task.getDeleted())
            {
                //Deleted tasks get processed then deleted from database
                //Modified time is not required but should be logged for potential bugs elsewhere
                if(!bModified)
                    Log.e(TAG,"ERROR, deleted task not modified");

                if(task.getHasTempId()) //If never added just delete from local database
                {
                    taskDb.delete(task);
                }
                else if (mAPI.tasks.delete(task))
                {
                    taskDb.delete(task);
                    dbToGoogle[SYNC_DELETE_TASK]++;
                    bListUpdated = true;
                }
            }
            else if(task.getHasTempId())
            {
                Task updated = mAPI.tasks.insert(task);
                if(updated != null) {
                    taskDb.delete(task);
                    task.setId(updated.getId());
                    task.setListId(updated.getListId());
                    taskDb.add(task);

                    dbToGoogle[SYNC_ADD_TASK]++;
                    bListUpdated = true;

                    if(listId.isEmpty()) //When adding a new task without a list this will be the default task list id
                        listId = updated.getListId();
                }
            }
            else if(bModified)
            {
                if(mAPI.tasks.update(task)) {
                    dbToGoogle[SYNC_CHANGE_TASK]++;
                    bListUpdated = true;
                }
            }

        }

        //If this function updated the list, need to retrieve it again to get new updated time
        if (bListUpdated) {
            TaskList updatedList = mAPI.taskLists.get(list.getId());
            if (updatedList != null) {
                Log.d(TAG, "New Updated = " + updatedList.getUpdated());
                webUpdated = updatedList.getUpdated(); //Updated modified time
            }
        }

        if(webUpdated.getTime() > savedUpdatedNEW.getTime()) {
            listDb.setLastUpdated(list.getId(), webUpdated);
        }
    }

    private static Task getTask(List<Task> tasks, String sId) {
        for(Task task : tasks) {
            if(task.getId().contentEquals(sId))
                return task;
        }

        return null;
    }
}

File: app/src/main/java/org/cerion/tasklist/sync/Sync.kt
package org.cerion.tasklist.sync


import android.content.Context
import android.util.Log
import org.cerion.tasklist.common.TAG
import org.cerion.tasklist.database.*
import org.cerion.tasklist.googleapi.GoogleApi
import org.cerion.tasklist.googleapi.GoogleTasksRepository
import java.util.*


internal class Sync(private val listDb: TaskListDao, private val taskDb: TaskDao, private val googleRepo: GoogleTasksRepository, private val prefs: Prefs) {

    fun lists(): SyncResult {
        // TODO add force sync flag if there was local changes never do this check
        if(Date().time - prefs.getDate(Prefs.KEY_LAST_LIST_SYNC).time < 1000 * 60) {
            Log.d(TAG,"Skipping lists since less than 1 minute")
            return SyncResult(true)
        }

        val result = SyncResult(false)

        val googleLists = googleRepo.getLists()
        if (googleLists.isEmpty())
            return result

        val dbLists = listDb.getAll()

        // Pair up default list Ids on first sync
        val defaultList = dbLists.getDefault()!!
        if (defaultList.hasTempId) {
            val googleId = googleLists.first { it.default }.id
            listDb.updateId(defaultList.id, googleId)
            defaultList.id = googleId
        }

        // Group lists in pair of <Google, Local>
        val map1 = googleLists.associateBy { it.id }
        val map2 = dbLists.associateBy { it.id }
        val pairs = (map1.keys + map2.keys).associateWith { Pair(map1[it], map2[it]) }.values

        pairs.forEach { pair ->
            val googleList = pair.first
            val localList = pair.second

            if (googleList != null && localList != null) {
                if (localList.deleted) {
                    val tasks = taskDb.getAllByList(localList.id)
                    if (tasks.isEmpty()) {
                        if (googleRepo.deleteList(localList)) {
                            listDb.delete(localList)
                            result.toRemote.delete++
                        }
                    }
                }
                else if (localList.updated != googleList.updated) {
                    var newUpdated = googleList.updated
                    if (localList.title != googleList.title) {
                        if (localList.updated > googleList.updated) {
                            newUpdated = googleRepo.updateList(localList).updated
                            result.toRemote.change++
                        }
                        else {
                            localList.title = googleList.title
                            result.toLocal.change++
                        }
                    }

                    localList.updated = newUpdated
                    listDb.update(localList)
                }
            }
            else if (googleList != null) {
                //Does not exist locally, add it
                listDb.add(TaskList(googleList.id, googleList.title).apply {
                    updated = googleList.updated
                })

                result.toLocal.add++
            }
            else if (localList != null) {
                if (localList.hasTempId) {
                    val addedList = googleRepo.createList(localList)

                    // If current displayed list is this one update the ID so it reloads with this list active
                    val lastId = prefs.getString(Prefs.KEY_LAST_SELECTED_LIST_ID)
                    if (lastId != null && lastId.contentEquals(localList.id))
                        prefs.setString(Prefs.KEY_LAST_SELECTED_LIST_ID, addedList.id)

                    // TODO added list will have a different updated time, that should be set here too
                    listDb.updateId(localList.id, addedList.id)
                    result.toRemote.add++
                }
                else {
                    // TODO this will cascade delete so log tasks count its removing...
                    listDb.delete(localList)
                    result.toLocal.delete++
                }
            }
            else
                throw UnsupportedOperationException() // should never happen
        }

        Log.d(TAG, "Lists to Local : ${result.toLocal.totalChanges} ${result.toLocal}")
        Log.d(TAG, "Lists to Google: ${result.toRemote.totalChanges} ${result.toRemote}")
        prefs.setDate(Prefs.KEY_LAST_LIST_SYNC, Date())

        result.success = true
        return result
    }

    fun tasks(listId: String): SyncResult {
        if (listId.isEmpty())
            return SyncResult(true)

        val result = SyncResult(false)

        val list = listDb.getAll().first { it.id == listId }
        val updatedMin = if(list.lastSync.time == 0L) null else list.lastSync
        val hasFullList = updatedMin == null
        val googleTasks = googleRepo.getTasks(listId, updatedMin)
        val localTasks = taskDb.getAllByList(listId)

        val map1 = googleTasks.associateBy { it.id }
        val map2 = localTasks.associateBy { it.id }
        val pairs = (map1.keys + map2.keys).associateWith { Pair(map1[it], map2[it]) }.values

        pairs.forEach { pair ->
            val googleTask = pair.first
            val localTask = pair.second

            if (googleTask != null && localTask != null) {
                if (googleTask.deleted) {
                    taskDb.delete(googleTask)
                    result.toLocal.delete++
                }
                else if (localTask.deleted) {
                    if (googleRepo.deleteTask(localTask)) {
                        taskDb.delete(localTask)
                        result.toRemote.delete++
                    }
                }
                else if (googleTask.updated != localTask.updated) {
                    var updateLocalTask = localTask
                    if (!googleTask.equals(localTask)) {
                        if (localTask.updated > googleTask.updated) {
                            updateLocalTask.updated = googleRepo.updateTask(localTask).updated
                            result.toRemote.change++
                        }
                        else {
                            updateLocalTask = googleTask
                            result.toLocal.change++
                        }
                    }
                    else { // no change but update last modified time
                        updateLocalTask.updated = googleTask.updated
                        Log.d(TAG, "No changes, updating last modified time")
                    }

                    taskDb.update(updateLocalTask)
                }
            }
            else if (googleTask != null) {
                if (googleTask.deleted) {
                    // Nothing, task already does not exist locally
                }
                else {
                    taskDb.add(googleTask)
                    result.toLocal.add++
                }
            }
            else if (localTask != null) {
                if (localTask.hasTempId && localTask.deleted) {
                    taskDb.delete(localTask)
                }
                else if (localTask.hasTempId) {
                    googleRepo.createTask(localTask)?.let { updated ->
                        taskDb.updateId(localTask.listId, localTask.id, updated.id, updated.updated)
                        result.toRemote.add++
                    }
                }
                else if (localTask.deleted) {
                    if (!hasFullList) {
                        googleRepo.deleteTask(localTask)
                        result.toRemote.delete++
                    }

                    taskDb.delete(localTask)
                }
                else if (hasFullList) { // Not in google, so must be deletion
                        result.toLocal.delete++
                        taskDb.delete(localTask)
                }
                else if (localTask.updated > list.lastSync) {
                    localTask.updated = googleRepo.updateTask(localTask).updated
                    taskDb.update(localTask)
                    result.toRemote.change++
                }
            }
        }

        Log.d(TAG, "Tasks to Local : ${result.toLocal.totalChanges} ${result.toLocal}")
        Log.d(TAG, "Tasks to Google: ${result.toRemote.totalChanges} ${result.toRemote}")

        // Update last sync time
        list.lastSync = Date()
        listDb.update(list)

        result.success = true
        return result
    }

    companion object {
        fun getInstance(context: Context, sAuthKey: String): Sync {
            val db = AppDatabase.getInstance(context)
            val api = GoogleApi(sAuthKey)
            val repo = GoogleTasksRepository(api)

            return Sync(db!!.taskListDao(), db.taskDao(), repo, Prefs.getInstance(context))
        }
    }
}

data class SyncResult(var success: Boolean) {
    val toLocal = SyncChanges()
    val toRemote = SyncChanges()

    val totalChanges: Int
        get() = toLocal.totalChanges + toRemote.totalChanges
}

data class SyncChanges(var add: Int = 0, var change: Int = 0, var delete: Int = 0) {
    val totalChanges: Int
        get() = add + change + delete
}